grpc-protocol/src/test/java/com/linecorp/armeria/common/grpc/protocol/Base64DecoderTest.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.grpc.protocol;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+
+import org.junit.jupiter.api.Test;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.base64.Base64;
+
+class Base64DecoderTest {
+
+    private static final ByteBuf[] EMPTY_BYTE_BUF = new ByteBuf[0];
+
+    @Test
+    void decodeConcatenatedBufsWithPadding() {
+        final String str = "abcd"; // YWJjZA==
+        final ByteBuf buf = Unpooled.wrappedBuffer(str.getBytes());
+        final ByteBuf encoded1 = Base64.encode(buf);
+        buf.readerIndex(0);
+        final ByteBuf encoded2 = Base64.encode(buf);
+        final ByteBuf concatenated = Unpooled.wrappedBuffer(encoded1, encoded2); // YWJjZA==YWJjZA==
+        final Base64Decoder base64Decoder = new Base64Decoder(PooledByteBufAllocator.DEFAULT);
+        final ByteBuf decoded = base64Decoder.decode(concatenated);
+        assertThat(decoded.toString(Charset.defaultCharset())).isEqualTo("abcdabcd");
+        decoded.release();
+    }
+
+    @Test
+    void decodeMultipleEncodedBytes() {
+        final String str = "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz1234567890";
+        final byte[] bytes = str.getBytes();
+        final List<ByteBuf> fragments = fragmentRandomly(bytes);
+        final int half = fragments.size() / 2;
+        final ByteBuf first =
+                Unpooled.wrappedBuffer(fragments.subList(0, half).toArray(EMPTY_BYTE_BUF));
+        final ByteBuf second =
+                Unpooled.wrappedBuffer(fragments.subList(half, fragments.size()).toArray(EMPTY_BYTE_BUF));
+        final Base64Decoder base64Decoder = new Base64Decoder(PooledByteBufAllocator.DEFAULT);
+        final ByteBuf decodedFirst = base64Decoder.decode(first);
+        final ByteBuf decodedSecond = base64Decoder.decode(second);
+        assertThat(Unpooled.wrappedBuffer(decodedFirst, decodedSecond).toString(Charset.defaultCharset()))
+                .isEqualTo(str);
+        decodedFirst.release();
+        decodedSecond.release();
+    }
+
+    private static List<ByteBuf> fragmentRandomly(byte[] bytes) {
```

Comment:
I don't think we should generate test cases randomly, it's very hard to be convinced this class is actually a complete test for base64, can you write a more concrete, easy to reason about test suite? Of course this is a big reason I didn't want to implement it :)

Reply:
I'm working on a PR that supports compression and decoding base64 in GrpcWebUtil.
Let me work on this in that PR. 😄